worker.go

```diff
@@ -20,16 +20,28 @@ func worker(context context.Context, waitGroup *sync.WaitGroup, firstTask func()
 	for {
 		select {
 		case <-context.Done():
-			// Pool context was cancelled, exit
+			// Pool context was cancelled, empty tasks channel and exit
+			for _ = range tasks {
+				taskWaitGroup.Done()
+			}
 			return
 		case task, ok := <-tasks:
-			if task == nil || !ok {
-				// We have received a signal to exit
-				return
-			}
+			// Prioritize context.Done statement (https://stackoverflow.com/questions/46200343/force-priority-of-go-select-statement)
+			select {
+			case <-context.Done():
+				if task != nil && ok {
+					// We have received a task, ignore it
+					taskWaitGroup.Done()
+				}
+			default:
+				if task == nil || !ok {
+					// We have received a signal to exit
+					return
+				}
 
-			// We have received a task, execute it
-			taskExecutor(task, false)
+				// We have received a task, execute it
+				taskExecutor(task, false)
```

Comment:
I spent some time reviewing this and I think we definitely need to prioritize context cancellation over grabbing tasks from the channel here. I propose a slightly different approach though, which should be equivalent to what you have here but that would prevent the 2nd select on the context after reading a task (inspiration from [this comment](https://www.reddit.com/r/golang/comments/1aoxwws/comment/kq2kd61/)):

```go
package pond

import (
	"context"
	"sync"
)

// worker represents a worker goroutine
func worker(context context.Context, waitGroup *sync.WaitGroup, firstTask func(), tasks <-chan func(), taskExecutor func(func(), bool), tasksWaitGroup *sync.WaitGroup) {

	// If provided, execute the first task immediately, before listening to the tasks channel
	if firstTask != nil {
		taskExecutor(firstTask, true)
	}

	defer func() {
		waitGroup.Done()
	}()

	for {
		// Prioritize context cancellation over grabbing pending tasks from the channel
		select {
		case <-context.Done():
			// Pool context was cancelled, empty tasks channel and exit
			drainTasks(tasks, tasksWaitGroup)
			return
		default:
		}

		select {
		case <-context.Done():
			// Pool context was cancelled, empty tasks channel and exit
			drainTasks(tasks, taskWaitGroup)
			return
		case task, ok := <-tasks:
			if task == nil || !ok {
				// We have received a signal to exit
				return
			}

			// We have received a task, execute it
			taskExecutor(task, false)
		}
	}
}

// drainPendingTasks discards queued tasks and decrements the corresponding wait group
func drainTasks(tasks <-chan func(), tasksWaitGroup *sync.WaitGroup) {
	for _ = range tasks {
		tasksWaitGroup.Done()
	}
}
```
Sorry I had to paste the entire file because github didn't let me to enter a suggestion in these lines :sweat_smile: 
The test you added was still passing with this change.

Reply:
With this version, in case both channels are set in the same time when we are in the second select (which would be the only one blocking), we could still execute the task instead of the clean up.

That's what they say in the first answer here https://stackoverflow.com/questions/46200343/force-priority-of-go-select-statement it se

It should be pretty rare in our case to get a new task in the same time as a done event but isn't it better to keep it that way to be sure it never happens?

Reply:
I think you are correct, there is a slight difference between the 2 versions. When both `contex.Done` and `tasks` channels have elements, the version I proposed would have a 50% chance of prioritizing the channel cancellation over the new task, but yours guarantees that channel cancellation is always honored over new tasks.